play-json/shared/src/main/scala-3/play/api/libs/json/JsMacroImpl.scala

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 2009-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package play.api.libs.json
+
+import scala.compiletime._
+import scala.deriving._
+
+/**
+ * Implementation for the JSON macro.
+ */
+object JsMacroImpl {
+  inline def format[A](using m: Mirror.Of[A]): OFormat[A] = OFormat[A](reads, writes)
+
+  inline def reads[A](using m: Mirror.Of[A]): Reads[A] = new Reads[A] { self =>
+    given subject: Reads[A] = this
+    def reads(js: JsValue)  = js match {
+      case obj @ JsObject(_) => inline m match {
+        case m: Mirror.ProductOf[A] => readElems[A](obj)(using self, m)
+        case m: Mirror.SumOf[A]     => readCases[A](obj)(using self, m)
+      }
+      case _                 => JsError("error.expected.jsobject")
+    }
+  }
+
+  inline def writes[A](using m: Mirror.Of[A]): OWrites[A] = new OWrites[A] { self =>
+    given subject: OWrites[A] = this
+    def writes(x: A)          = inline m match {
+      case m: Mirror.ProductOf[A] => writeElems[A](x)(using self, m)
+      case m: Mirror.SumOf[A]     => writeCases[A](x)(using self, m)
+    }
+  }
+
+  inline def readElems[A: Reads](obj: JsObject)(using m: Mirror.ProductOf[A]): JsResult[A] = {
+    inline val size = constValue[Tuple.Size[m.MirroredElemTypes]]
+    readElemsL[A, m.MirroredElemLabels, m.MirroredElemTypes](obj, new Array[Any](size))(0)
+  }
+
+  inline def writeElems[A: OWrites](x: A)(using m: Mirror.ProductOf[A]): JsObject =
+    writeElemsL[A, m.MirroredElemLabels, m.MirroredElemTypes](x)(0, JsObject.empty)
+
+  inline def readCases[A: Reads](obj: JsObject)(using m: Mirror.SumOf[A]): JsResult[A] = {
+    obj.value.get(config.discriminator) match {
+      case None      => JsError(JsPath \ config.discriminator, "error.missing.path")
+      case Some(tjs) => {
+        val vjs = obj.value.get("_value").getOrElse(obj)
+        tjs.validate[String].flatMap { dis =>
+          // lampepfl/dotty#11048 discriminator may be "play.api.libs.json.MacroSpec.Simple" but in MirroredElemLabels is "Simple"
+          val disName = dis.drop(dis.lastIndexOf('.') + 1)
+          readCasesL[A, m.MirroredElemLabels, m.MirroredElemTypes](vjs, disName)
+        }
+      }
+    }
+  }
+
+  inline def writeCases[A: OWrites](x: A)(using m: Mirror.SumOf[A]): JsObject =
+    writeCasesL[A, m.MirroredElemLabels, m.MirroredElemTypes](x, m.ordinal(x))(0)
+
+  inline def readElemsL[A: Reads, L <: Tuple, T <: Tuple](obj: JsObject, elems: Array[Any])(n: Int)(using m: Mirror.ProductOf[A]): JsResult[A] =
+    inline (erasedValue[L], erasedValue[T]) match {
+      case _: (EmptyTuple, EmptyTuple) => JsSuccess(m.fromProduct(new ArrayProduct(elems)))
```

Comment:
we could use TupleXXL instead of introducing a custom ArrayProduct
```suggestion
      case _: (EmptyTuple, EmptyTuple) => JsSuccess(m.fromProduct(scala.runtime.TupleXXL.fromIArray(elems.asInstanceOf[IArray[Object]])))
```

Reply:
Nice find that TupleXXL fits.  But scala.runtime is the part of the standard library which is private to Scala - i.e. it's not stable public API.  So I think it's best to use our own little subclass.

Reply:
I realized we can use `Tuple.fromIArray` or even `Tuple.fromArray`, but there are some indirections there https://github.com/lampepfl/dotty/blob/master/library/src/scala/runtime/Tuples.scala

Reply:
Yeah, true...  No point in allocating another tuple.